x/exchange/keeper/grpc_query.go

```diff
@@ -545,3 +545,132 @@ func (k QueryServer) ValidateManageFees(goCtx context.Context, req *exchange.Que
 
 	return resp, nil
 }
+
+// GetPayment gets a single specific payment.
+func (k QueryServer) GetPayment(goCtx context.Context, req *exchange.QueryGetPaymentRequest) (*exchange.QueryGetPaymentResponse, error) {
+	if req == nil || len(req.Source) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	source, err := sdk.AccAddressFromBech32(req.Source)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid source %q: %v", req.Source, err)
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	resp := &exchange.QueryGetPaymentResponse{}
+	resp.Payment, _ = k.Keeper.GetPayment(ctx, source, req.ExternalId)
+	if resp.Payment == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "no payment found with source %s and external id %q",
+			req.Source, req.ExternalId)
+	}
+
+	return resp, nil
+}
+
+// GetPaymentsWithSource gets all payments with a specific source account.
+func (k QueryServer) GetPaymentsWithSource(goCtx context.Context, req *exchange.QueryGetPaymentsWithSourceRequest) (*exchange.QueryGetPaymentsWithSourceResponse, error) {
+	if req == nil || len(req.Source) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	source, err := sdk.AccAddressFromBech32(req.Source)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid source %q: %v", req.Source, err)
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	keyPrefix := GetKeyPrefixPaymentsForSource(source)
+	preStore := prefix.NewStore(k.getStore(ctx), keyPrefix)
+
+	resp := &exchange.QueryGetPaymentsWithSourceResponse{}
+	var pageErr error
+	resp.Pagination, pageErr = query.Paginate(preStore, req.Pagination, func(_, value []byte) error {
+		// Only add it to the result if we can read it. This might result in fewer results than the limit,
+		// but at least one bad entry won't block others by causing the whole thing to return an error.
+		payment, _ := k.parsePaymentStoreValue(value)
```

Comment:
One minor concern with this approach is it will mask incorrect values. I would recommend at least logging the error so we know something is incorrect in the state.

Reply:
Yeah. Good call. I added logging to these queries. I also updated `GetPayment` to return an error if there's an error reading the payment.